File: VB/EditorsDemo.Wpf/Modules/LookUpEditServerMode/LookUpInstantFeedbackModeViewModel.vb
Imports DevExpress.Data.Linq
Imports DevExpress.Xpf.DemoBase
Imports System.Collections.Generic
Imports System.Data.SQLite

Namespace GridDemo
    Public Class LookUpInstantFeedbackModeViewModel
        Inherits InstantFeedbackModeViewModelBase(Of Person)

        #Region "DatabaseInfo"
        Private Shared ReadOnly Generator As New PersonGenerator()
        Public Shared ReadOnly DatabaseInfo As New DatabaseInfo(PersonsDataContext.FileName, "People", GetType(Person), AddressOf CreateValues, Function(sql, connection) New SQLiteCommand(sql, CType(connection, SQLiteConnection)))
        Private Shared Function CreateValues() As IDictionary(Of String, Object)
            Dim person = Generator.CreatePerson()
            Dim dict = New Dictionary(Of String, Object)()
            dict.Add("FullName", person.FullName)
            dict.Add("Company", person.Company)
            dict.Add("JobTitle", person.JobTitle)
            dict.Add("City", person.City)
            dict.Add("Address", person.Address)
            dict.Add("Phone", person.Phone)
            dict.Add("Email", person.Email)
            Return dict
        End Function
        #End Region

        Protected Sub New()
            MyBase.New(DatabaseInfo, Function() (New PersonsDataContext()).People)
            Orders = OrderDataGenerator.GenerateOrders(200)
        End Sub

        Private privateOrders As List(Of OrderData)
        Public Property Orders() As List(Of OrderData)
            Get
                Return privateOrders
            End Get
            Private Set(ByVal value As List(Of OrderData))
                privateOrders = value
            End Set
        End Property
        Public Overridable Property InstantFeedbackSource() As EntityInstantFeedbackSource

        Protected Overrides Sub AssignSourcesCore()
            MyBase.AssignSourcesCore()
            DisposeInstantFeedbackSource()
            Dim source = New EntityInstantFeedbackSource() With {.KeyExpression = "Id"}
            AddHandler source.GetQueryable, Sub(o, e) e.QueryableSource = getQueryable()
            InstantFeedbackSource = source
        End Sub
        Public Overrides Sub OnUnloaded()
            MyBase.OnUnloaded()
            DisposeInstantFeedbackSource()
        End Sub

        Private Sub DisposeInstantFeedbackSource()
            If InstantFeedbackSource IsNot Nothing Then
                InstantFeedbackSource.Dispose()
                InstantFeedbackSource = Nothing
            End If
        End Sub
    End Class
End Namespace

File: CS/GridDemo.Wpf/Modules/LookUpEditServerMode/LookUpInstantFeedbackModeViewModel.cs
using DevExpress.Data.Linq;
using DevExpress.Xpf.DemoBase;
using System.Collections.Generic;
using System.Data.SQLite;

namespace GridDemo {
    public class LookUpInstantFeedbackModeViewModel : InstantFeedbackModeViewModelBase<Person> {
        #region DatabaseInfo
        static readonly PersonGenerator Generator = new PersonGenerator();
        public static readonly DatabaseInfo DatabaseInfo = new DatabaseInfo(
            PersonsDataContext.FileName, "People", typeof(Person), CreateValues, (sql, connection) => new SQLiteCommand(sql, (SQLiteConnection)connection));
        static IDictionary<string, object> CreateValues() {
            var person = Generator.CreatePerson();
            var dict = new Dictionary<string, object>();
            dict.Add("FullName", person.FullName);
            dict.Add("Company", person.Company);
            dict.Add("JobTitle", person.JobTitle);
            dict.Add("City", person.City);
            dict.Add("Address", person.Address);
            dict.Add("Phone", person.Phone);
            dict.Add("Email", person.Email);
            return dict;
        } 
        #endregion

        protected LookUpInstantFeedbackModeViewModel()
            : base(DatabaseInfo, () => new PersonsDataContext().People) {
            Orders = OrderDataGenerator.GenerateOrders(200);
        }

        public List<OrderData> Orders { get; private set; }
        public virtual EntityInstantFeedbackSource InstantFeedbackSource { get; set; }

        protected override void AssignSourcesCore() {
            base.AssignSourcesCore();
            DisposeInstantFeedbackSource();
            var source = new EntityInstantFeedbackSource() {
                KeyExpression = "Id",
            };
            source.GetQueryable += (o, e) => {
                e.QueryableSource = getQueryable();
            };
            InstantFeedbackSource = source;
        }
        public override void OnUnloaded() {
            base.OnUnloaded();
            DisposeInstantFeedbackSource();
        }

        void DisposeInstantFeedbackSource() {
            if(InstantFeedbackSource != null) {
                InstantFeedbackSource.Dispose();
                InstantFeedbackSource = null;
            }
        }
    }
}
